iotilecore/iotile/core/hw/debug/firmware_image_analyzer.py

```diff
@@ -0,0 +1,140 @@
+        self.num_addresses = len(self._hex_image.addresses())
+        self.segments = self._hex_image.segments()
+        self.memory_size = self._hex_image.get_memory_size()
+        self.cdb_app_info = self.get_cdb_app_info()
+        if self.cdb_app_info is None:
+            raise DataError("Unable to analyze firmware image.")
+
+    def get_firmware_hardware_type(self):
+        """Returns hardware type specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return self.cdb_app_info['hardware_type']
+
+    def get_firmware_api_version(self):
+        """Returns firmware api version specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return (self.cdb_app_info['api_major_version'],
+                self.cdb_app_info['api_minor_version'])
+
+    def get_firmware_name(self):
+        """Returns firmware name specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return self.cdb_app_info['name'].decode("utf-8")
+
+    def get_firmware_version(self):
+        """Returns firmware module version specified in cdb app info"""
+        if self.cdb_app_info is None:
+            self.get_cdb_app_info()
+
+        return (self.cdb_app_info['module_major_version'],
+                self.cdb_app_info['module_minor_version'],
+                self.cdb_app_info['module_patch_version'])
+
+    def get_cdb_app_info(self):
+        """Gets the string of bytes of the CDB app info"""
+        cdb_app_info_bytes = self._hex_image.gets(self.max_addr - (FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO - 1), \
+                FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO)
+        if self._check_cdb_app_info(cdb_app_info_bytes):
+            self.cdb_app_info = format_cdb_app_info(cdb_app_info_bytes)
+            return self.cdb_app_info
+
+        addresses = self._hex_image.addresses()
+        for address_index in range(0, self.num_addresses - 4):
+            search_bytes = self._hex_image.gets(addresses[address_index], 4)
+            search, = struct.unpack("<L", search_bytes)
+            if search == self.magic_number:
+                cdb_app_info_bytes = self._hex_image.gets(addresses[address_index - FirmwareImageAnalyzer.MAGIC_NUMBER_BYTE_OFFSET], \
+                    FirmwareImageAnalyzer.SIZE_OF_CDBAPPINFO)
+                if self._check_cdb_app_info(cdb_app_info_bytes):
+                    self.cdb_app_info = format_cdb_app_info(cdb_app_info_bytes)
+                    return self.cdb_app_info
+
+        return None
+
+    def _check_cdb_app_info(self, cdb_app_info_bytes):
+        """Helper function to check the byte string if it is a valid CDB app info"""
+        cdb_app_info = format_cdb_app_info(cdb_app_info_bytes)
+
+        try:
+            cdb_app_info['name'].decode("utf-8")
+        except UnicodeDecodeError:
+            return False
+        except:
```

Comment:
You still need to justify this exception. Why are you raising an ExternalError instead of just letting the original exception raise? Is the caller expected to handle an `ExternalError` instead of whatever does happen? 

If that's the case, you need to put a `pylint=disable-bare-except` clause on this to supress the pylint warning. 

Unless we never want to fix the cause of the ExternalError, we shouldn't be overwriting what the actual error is. If you *do* need to raise it instead of the original one, you should at least log the exception and stacktrace.

Reply:
One option is:

```
except Exception as err:
    raise ExternalError("Unknown whatever") from err
```

That fully preserves the original exception information and stacktrace so it can be debugged.

Reply:
Added Tim's implementation.